nengo_fpga/dna_extractor.py

```diff
@@ -0,0 +1,258 @@
+import socket
+import os
+import threading
+import paramiko
+import numpy as np
+
+from nengo_fpga.fpga_config import fpga_config
+
+
+class DNAExtractor(object):
+    """ Class that connects to the FPGA and extracts the Device DNA
+
+    Parameters
+    ----------
+    fpga_name : str
+        The name of the fpga defined in the config file.
+    max_attempts : int, optional (Default: 5)
+        The number of times the socket will attempt to connect to the board.
+    timeout : float, optional (Default: 5)
+        The number of seconds the socket will wait to connect.
+
+    """
+    def __init__(self, fpga_name, max_attempts=5, timeout=5):
+
+        self.config_found = fpga_config.has_section(fpga_name)
+        self.fpga_name = fpga_name
+        self.max_attempts = max_attempts
+        self.timeout = timeout
+
+        # Make SSHClient object
+        self.ssh_client = paramiko.SSHClient()
+        self.ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        self.ssh_info_str = ''
+        self.ssh_lock = False
+
+        # Check if the desired FPGA name is defined in the configuration file
+        if self.config_found:
+            # Handle the tcp port selection: Use the config specified port.
+            # If none is provided (i.e., the specified port number is 0),
+            # choose a random tcp port number between 20000 and 65535.
+            # We will use the udp port number from the config but use tcp.
+            self.tcp_port = int(fpga_config.get(fpga_name, 'udp_port'))
+            if self.tcp_port == 0:
+                self.tcp_port = int(np.random.uniform(low=20000, high=65535))
+
+            # Make the TCP socket for receiving DNA.
+            self.tcp_init = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            self.tcp_init.bind((fpga_config.get('host', 'ip'), self.tcp_port))
+            self.tcp_init.settimeout(self.timeout)
+            self.tcp_init.listen(1)  # Ready to accept a connection
+
+        else:
+            # FPGA name not found, throw a warning.
+            print("WARNING: Specified FPGA configuration '" + fpga_name +
+                  "' not found.", flush=True)
+            self.tcp_socket = None
+
+    def cleanup(self):
+        self.tcp_init.close()
+        self.tcp_recv.close()
```

Comment:
If the connection attempts time out, `self.tcp_recv` is not a valid object (because it never gets created), so this line will fail. We should put a check here:

```
if self.tcp_recv is not None:
    self.tcp_recv.close()
```

and then make the appropriate changes in the class constructor.